driver/ppm_fillers.c

```diff
@@ -7951,5 +7951,32 @@ int f_sys_memfd_create_x(struct event_filler_arguments *args)
 	res = val_to_ring(args, memfd_create_flags_to_scap(val), 0, true, 0);
 	CHECK_RES(res);
 
+	return add_sentinel(args);
+}
+int f_sys_pidfd_getfd_x(struct event_filler_arguments *args){
+	unsigned long val;
+	int res;
+	int retval;
+
+	/* Parameter 1: ret (type: PT_FD) */
+	retval = (int64_t) syscall_get_return_value(current,args->regs);
+	res = val_to_ring(args, retval, 0, false, 0);
+	CHECK_RES(res)
+	
+	/* Parameter 2: pidfd (type: PT_FD) */
+	syscall_get_arguments_deprecated(args, 0, 1, &val);
```

Comment:
@Andreagit97, I keep getting stuck or unaware of when why and where we must use type conversions. I'm aware of this https://github.com/falcosecurity/libs/pull/783#issue-1486570992 but can't quite understand it. Could you please explain briefly or link and article for me!!

Reply:
sure!
As you see in your tests, the `pid_fd` syscall argument is 32-bit and signed, so in the kernel, we should expect to retrieve a signed number on 32-bit as well. `pid_fd` is  `-1` on 32 bit and it will be represented as `0xFFFFFFFF`.

```cpp
    int pid_fd = -1;
    int target_fd = -1;
    uint32_t flags = 1;
    assert_syscall_state(SYSCALL_FAILURE, "pidfd_getfd", syscall(__NR_pidfd_getfd, pid_fd, target_fd, flags));
```

Let's see the kernel side considering the bpf probe as an example.
The `bpf_syscall_get_argument` method will return us an unsigned long so a number on 64-bit. In the specific case of `pid_fd` it will return us `0x00000000FFFFFFFF`, the high 32 bits contain all 0 while the lower ones contain the real number we want: `-1`. If we send this unsigned long as it is to userspace we will obtain something weird.

```c
/* Parameter 2: pidfd (type: PT_FD) */
unsigned long pidfd = bpf_syscall_get_argument(data, 0);
res = bpf_push_s64_to_ring(data, pidfd);
CHECK_RES(res);
```

We are sending to userspace an unsigned long `0x00000000FFFFFFFF` telling to treat it as an `s64` so a signed number on 64 bit. Userspace will read `4294967295` instead of `-1` because `-1` on 64-bit should be represented as `0xFFFFFFFFFFFFFFFF` and not as `0x00000000FFFFFFFF`. What we have to do in order to allow the userspace to understand is:
 
1. Keep only the first 32 bits of the `bpf_syscall_get_argument` method:

```c
s32 pidfd = (s32)bpf_syscall_get_argument(data, 0);
```
   Now we have s32 pidfd = `0xFFFFFFFF`. 

2. Before sending it we need to convert it to a signed 64-bit and not to an unsigned long:


```c
res = bpf_push_s64_to_ring(data, (s64)pidfd)
```

with this  `(s64)pidfd` we will convert `0xFFFFFFFF` to `0xFFFFFFFFFFFFFFFF`. 

Now the userspace can correctly read `-1`:tada:

Reply:
This makes sense!! Thank you for the detailed explanation!!